src/include/codegen/query.h

```diff
@@ -55,29 +57,48 @@ class Query {
 
   // Execute th e query given the catalog manager and runtime/consumer state
   // that is passed along to the query execution code.
-  void Execute(concurrency::Transaction &txn,
-               executor::ExecutorContext *executor_context, char *consumer_arg,
+  void Execute(executor::ExecutorContext &executor_context,
+               QueryParameters &parameters, char *consumer_arg,
                RuntimeStats *stats = nullptr);
 
   // Return the query plan
   const planner::AbstractPlan &GetPlan() const { return query_plan_; }
 
+  // Return the query parameters
+  codegen::QueryParameters &GetQueryParameters() const {
+    return parameters_;
+  }
+
   // Get the holder of the code
   CodeContext &GetCodeContext() { return code_context_; }
 
   // The class tracking all the state needed by this query
   RuntimeState &GetRuntimeState() { return runtime_state_; }
 
+  size_t GetParameterIdx(const expression::AbstractExpression *expression) {
+    return parameters_.GetParameterIdx(expression);
+  }
+
+  ParameterCache &GetParameterCache() {
+    return parameters_cache_;
+  }
+
  private:
   friend class QueryCompiler;
 
   // Constructor
-  Query(const planner::AbstractPlan &query_plan);
+  Query(const planner::AbstractPlan &query_plan, QueryParameters &parameters);
```

Comment:
First,I agree that `Query` does not have to (or should not) carry `QueryParameters` and `ParameterCache`. It was something that I missed to refactor after building a working version. I moved them to `CompilationContext`. 

For the second part of your question, I think the separation of those concepts is not much useful for `CompilationContext`.  This is because the parameter cache in the `CompilationContext` should know about the _values_(types, nullability) of the parameters, although the values are not required.  The separation of actual values from the types would clarify this situation.  I will give another thought on this while fixing the null, below, which I thought IsNull() will be incorrect..

Thanks for point these out.